File: code/sorting/shaker_sort/shaker_sort.rs
/* Part of Cosmos by OpenGenus Foundation */
/* Added by Ondama */

/** Read a line of input **/
fn read () -> String {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).expect("failed to read from stdin");
    input.trim().to_owned()
}

/** Display a message and parse input **/
fn prompt<T: std::str::FromStr + std::default::Default> (prompt: &str) -> T {
    println!("{}", prompt);
    let num = read();
    match num.parse() {
        Ok(i) => i,
        Err(..) => {
            eprintln!("'{}' is not valid, using default instead.", num);
            T::default()
        }
    }
}

/** Sort array **/
fn shaker_sort<T: std::cmp::PartialOrd + std::fmt::Debug> (arr: &mut Vec<T>) {
    let size = arr.len();
    for i in 0..size/2 {
        let mut swapped: bool = false;
        let k = size - i - 1;
        for j in i..k {
            let p = j;
            let q = p + 1;
            if arr[p] < arr[q] {
                arr.swap(p, q);
                swapped = true;
            }
        }
        for j in i..k {
            let p = k - j;
            let q = p - 1;
            if arr[p] > arr[q] {
                arr.swap(p, q);
                swapped = true;
            }
        }
        if !swapped {
            break;
        }
    }
}

/** Entry point **/
fn main () {
    // The size is a natural number
    let size = prompt::<usize>(&"Enter size of array:");
    // Fill array
    let mut arr = Vec::new();
    loop {
        let item = prompt::<i32>(&"Next array element:");
        arr.push(item);
        if arr.len() == size {
            break;
        }
    }
    // Sort array
    shaker_sort::<i32>(&mut arr);
    // Print sorted array
    println!("{:?}", arr);
}

File: code/sorting/src/shaker_sort/shaker_sort.php
<?php

/* Part of Cosmos by OpenGenus Foundation */

namespace OpenGenus;

trait ShakerSortTrait {
    /**
     * @param array $list
     * @return array
     */
    public function sort(array $list) : array
    {
        do {
            $swapped = false;
            $length = count($list);

            for ($i = 0; $i < $length - 2; $i++) {
                if ($list[$i] > $list[$i + 1]) {
                    $list = $this->swap($list, $i, $i + 1);
                    $swapped = true;
                }
            }

            if (false === $swapped) {
                break;
            }

            $swapped = false;
            for ($i = $length - 2; $i > 0; $i--) {
                if ($list[$i] > $list[$i + 1]) {
                    $list = $this->swap($list, $i, $i + 1);
                    $swapped = true;
                }
            }
        } while(true === $swapped);

        return $list;
    }

    /**
     * @param array $input
     * @param int $fromIndex
     * @param int $toIndex
     * @return array
     */
    private function swap(array $input, int $fromIndex, int $toIndex) : array
    {
        $swapFrom = $input[$fromIndex];
        $input[$fromIndex] = $input[$toIndex];
        $input[$toIndex] = $swapFrom;

        return $input;
    }
}